pkg/node/fsm/wait_micro_snapshot.go

```diff
@@ -0,0 +1,244 @@
+				"failed to get snapshot for microBlock '%s' - timeout", t.BlockID))
+		default:
+			return a, nil, a.Errorf(errors.New("undefined Snapshot Task type"))
+		}
+	default:
+		return a, nil, a.Errorf(errors.Errorf(
+			"unexpected internal task '%d' with data '%+v' received by %s State",
+			task.TaskType, task.Data, a.String()))
+	}
+}
+
+func (a *WaitMicroSnapshotState) MicroBlockSnapshot(
+	_ peer.Peer,
+	blockID proto.BlockID,
+	snapshot proto.BlockSnapshot,
+) (State, Async, error) {
+	if a.microBlockWaitingForSnapshot == nil {
+		return a, nil, nil
+	}
+	if a.microBlockWaitingForSnapshot.TotalBlockID != blockID {
+		return a, nil, a.Errorf(errors.Errorf(
+			"New snapshot doesn't match with microBlock %s", a.microBlockWaitingForSnapshot.TotalBlockID))
+	}
+	// the TopBlock() is used here
+	block, err := a.checkAndAppendMicroBlock(a.microBlockWaitingForSnapshot, &snapshot)
+	if err != nil {
+		metrics.FSMMicroBlockDeclined("ng", a.microBlockWaitingForSnapshot, err)
+		return a, nil, a.Errorf(err)
+	}
+	defer func() {
+		a.microBlockWaitingForSnapshot = nil
+		close(a.timeoutTaskOutdated)
+		a.timeoutTaskOutdated = nil
+	}()
+	zap.S().Named(logging.FSMNamespace).Debugf(
+		"[%s] Received snapshot for microblock '%s' successfully applied to state", a, block.BlockID(),
+	)
+	a.baseInfo.MicroBlockCache.AddMicroBlockWithSnapshot(block.BlockID(), a.microBlockWaitingForSnapshot, &snapshot)
+	a.blocksCache.AddBlockState(block)
+	a.blocksCache.AddSnapshot(block.BlockID(), snapshot)
+	a.baseInfo.scheduler.Reschedule()
+	// Notify all connected peers about new microblock, send them microblock inv network message
+	if inv, ok := a.baseInfo.MicroBlockInvCache.Get(block.BlockID()); ok {
+		//TODO: We have to exclude from recipients peers that already have this microblock
+		if err = broadcastMicroBlockInv(a.baseInfo, inv); err != nil {
+			return a, nil, a.Errorf(errors.Wrap(err, "failed to handle microblock message"))
+		}
+	}
+	return a, nil, nil
+}
+
+func (a *WaitMicroSnapshotState) checkAndAppendMicroBlock(
+	micro *proto.MicroBlock,
+	snapshot *proto.BlockSnapshot,
+) (*proto.Block, error) {
+	top := a.baseInfo.storage.TopBlock()  // Get the last block
+	if top.BlockID() != micro.Reference { // Microblock doesn't refer to last block
+		err := errors.Errorf("microblock TBID '%s' refer to block ID '%s' but last block ID is '%s'",
+			micro.TotalBlockID.String(), micro.Reference.String(), top.BlockID().String())
+		metrics.FSMMicroBlockDeclined("ng", micro, err)
+		return &proto.Block{}, proto.NewInfoMsg(err)
+	}
+	ok, err := micro.VerifySignature(a.baseInfo.scheme)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, errors.Errorf("microblock '%s' has invalid signature", micro.TotalBlockID.String())
+	}
+	newTrs := top.Transactions.Join(micro.Transactions)
+	newBlock, err := proto.CreateBlock(newTrs, top.Timestamp, top.Parent, top.GeneratorPublicKey, top.NxtConsensus,
+		top.Version, top.Features, top.RewardVote, a.baseInfo.scheme)
```

Comment:
New block should contain `StateHash` field, which should be checked. 
`StateHash` field must be initialized after light node feature activation + 1000 blocks, before that the filed must not be initialized. This check depends on #1254, so for now passing & calculatiing StateHash when its passed will be enough.

Reply:
@Anton-Rampage A new PR is required for resolving this issue.